brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java

```diff
@@ -0,0 +1,76 @@
+package brave.internal.propagation;
+
+import brave.internal.HexCodec;
+import brave.internal.Nullable;
+import brave.propagation.CurrentTraceContext.Scope;
+import brave.propagation.CurrentTraceContext.ScopeDecorator;
+import brave.propagation.TraceContext;
+
+import static brave.internal.HexCodec.lowerHexEqualsTraceId;
+import static brave.internal.HexCodec.lowerHexEqualsUnsignedLong;
+
+/**
+ * Adds correlation properties "traceId", "parentId" and "spanId" when a {@link
+ * brave.Tracer#currentSpan() span is current}.
+ */
+public abstract class CorrelationFieldScopeDecorator implements ScopeDecorator {
+
+  @Override public Scope decorateScope(@Nullable TraceContext currentSpan, Scope scope) {
+    String previousTraceId = getIfString("traceId");
+    String previousSpanId = getIfString("spanId");
+    String previousParentId = getIfString("parentId");
+
+    if (currentSpan != null) {
+      maybeReplaceTraceContext(currentSpan, previousTraceId, previousParentId, previousSpanId);
+    } else {
+      remove("traceId");
+      remove("parentId");
+      remove("spanId");
+    }
+
+    class CorrelationFieldCurrentTraceContextScope implements Scope {
+      @Override public void close() {
+        scope.close();
+        replace("traceId", previousTraceId);
+        replace("parentId", previousParentId);
+        replace("spanId", previousSpanId);
+      }
+    }
+    return new CorrelationFieldCurrentTraceContextScope();
+  }
+
+  void maybeReplaceTraceContext(
+      TraceContext currentSpan,
+      String previousTraceId,
+      @Nullable String previousParentId,
+      String previousSpanId
+  ) {
+    boolean sameTraceId = lowerHexEqualsTraceId(previousTraceId, currentSpan);
+    if (!sameTraceId) put("traceId", currentSpan.traceIdString());
+
+    long parentId = currentSpan.parentIdAsLong();
+    if (parentId == 0L) {
+      remove("parentId");
+    } else {
+      boolean sameParentId = lowerHexEqualsUnsignedLong(previousParentId, parentId);
+      if (!sameParentId) put("parentId", HexCodec.toLowerHex(parentId));
+    }
+
+    boolean sameSpanId = lowerHexEqualsUnsignedLong(previousSpanId, currentSpan.spanId());
+    if (!sameSpanId) put("spanId", HexCodec.toLowerHex(currentSpan.spanId()));
+  }
+
+  protected abstract @Nullable String getIfString(String key);
```

Comment:
Javadoc? Also I'm not sure how to grok this `get` - at first I expected it to get the value from the current trace context. `getPreviouslyCorrelated` (a bit wordy though)?

Reply:
yep. it is internal, but no excuse :P